doc/adr/0006-protocol-naming-conventions.md

```diff
@@ -0,0 +1,56 @@
+# 6. protocol-naming-conventions
+
+Date: 2021-05-08
+
+## Status
+
+Under Review
+
+## Context
+
+This document describes naming conventions for these protocol components:
+
+* message subject
+* stream name
+* reply to
+* queue name
+* durable name
+* account name
+
+## Prior Work
+
+Currently the NATS Docs regarding [protocol convention](https://docs.nats.io/nats-protocol/nats-protocol#protocol-conventions) says this:
+
+> Subject names, including reply subject (INBOX) names, are case-sensitive and must be non-empty alphanumeric strings with no embedded whitespace. All ascii alphanumeric characters except spaces/tabs and separators which are "." and ">" are allowed. Subject names can be optionally token-delimited using the dot character (.), e.g.:
+A subject is comprised of 1 or more tokens. Tokens are separated by "." and can be any non space ascii alphanumeric character. The full wildcard token ">" is only valid as the last token and matches all tokens past that point. A token wildcard, "*" matches any token in the position it was listed. Wildcard tokens should only be used in a wildcard capacity and not part of a literal token.
+
+> Character Encoding: Subject names should be ascii characters for maximum interoperability. Due to language constraints and performance, some clients may support UTF-8 subject names, as may the server. No guarantees of non-ASCII support are provided.
+
+## Specification
+
+```
+dot              = "."
+asterisk         = "*"
+lt               = "<"
+gt               = ">"
+dollar           = "$"
+colon            = ":"
+double-quote     = ["]
+fwd-slash        = "/"
+backslash        = "\"
+pipe             = "|"
+question-mark    = "?"
+ampersand        = "&"
+printable        = all printable ascii (33 to 126 inclusive)
+term             = (printable except dot, asterisk or gt)+
+prefix           = (printable except dot, asterisk, gt or dollar)+
```

Comment:
Is the `dollar` in the except list? Yet the default JS prefix is `$JS.API.`

Reply:
changed

Reply:
right, `dollar` can be in a server / internal prefix but not a user prefix. (user prefix changed here too). Maybe leave prefix and add this?
```
internal-prefix = dollar (prefix dot)+
user-prefix     = (prefix dot)+
```

Reply:
Sure. And to be clear: by prefix you mean the JS API prefix, right? Because stream names allow "$" and MQTT implementation in the server creates streams that start with `$MQTT_`, so wanted to be sure that this is different.

Reply:
correct the JS api prefix. I will change the
```
jetstream-prefix = prefix
```
to
```
js-internal-prefix = dollar (prefix dot)+
js-user-prefix     = (prefix dot)+
```